File: examples/agent-framework-webhooks/python/agent_webhooks/adapters/langchain.py
"""LangChain 1.x reply adapter."""

import os
from collections.abc import Awaitable, Callable, Mapping, Sequence
from typing import Any

from e2a import AsyncInboundEmail

from agent_webhooks.prompt import REPLY_INSTRUCTIONS, email_prompt

LangChainRun = Callable[[str], Awaitable[Mapping[str, Any]]]


def _field(value: Any, name: str) -> Any:
    if isinstance(value, Mapping):
        return value.get(name)
    return getattr(value, name, None)


def _text_content(content: Any) -> str:
    if isinstance(content, str):
        return content
    if not isinstance(content, Sequence):
        return ""
    return "\n".join(
        text
        for block in content
        if _field(block, "type") == "text"
        and isinstance((text := _field(block, "text")), str)
    )


class LangChainReplyAgent:
    def __init__(self, run: LangChainRun) -> None:
        self._run = run

    @classmethod
    def from_env(cls) -> "LangChainReplyAgent":
        from langchain.agents import create_agent

        agent = create_agent(
            model=os.getenv("LANGCHAIN_MODEL", "openai:gpt-5.5"),
            tools=[],
            system_prompt=REPLY_INSTRUCTIONS,
        )
        return cls(
            lambda prompt: agent.ainvoke(
                {"messages": [{"role": "user", "content": prompt}]}
            )
        )

    async def reply(
        self, email: AsyncInboundEmail, conversation_id: str
    ) -> str:
        del conversation_id
        result = await self._run(email_prompt(email))
        messages = result.get("messages", [])
        for message in reversed(messages):
            if (
                _field(message, "type") == "ai"
                or _field(message, "role") == "assistant"
            ):
                return _text_content(_field(message, "content"))
        raise ValueError("LangChain result did not contain an assistant message")

File: examples/agent-framework-webhooks/typescript/src/adapters/langchain.ts
import type { InboundEmail } from "@e2a/sdk/v1";
import { createAgent } from "langchain";

import type { ReplyAgent } from "../contracts.js";
import { REPLY_INSTRUCTIONS, emailPrompt } from "../prompt.js";

export interface LangChainResult {
  messages?: readonly unknown[];
}

export type LangChainRun = (prompt: string) => Promise<LangChainResult>;

function field(value: unknown, name: "type" | "role" | "content" | "text"): unknown {
  if (!value || typeof value !== "object") return undefined;
  return (value as Record<string, unknown>)[name];
}

function textContent(content: unknown): string {
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";
  return content.flatMap((block) => {
    const text = field(block, "text");
    return field(block, "type") === "text" && typeof text === "string" ? [text] : [];
  }).join("\n");
}

export class LangChainReplyAgent implements ReplyAgent {
  constructor(private readonly run: LangChainRun) {}

  async reply(email: InboundEmail, _conversationId: string): Promise<string> {
    const result = await this.run(emailPrompt(email));
    for (const message of [...(result.messages ?? [])].reverse()) {
      if (field(message, "type") === "ai" || field(message, "role") === "assistant") {
        return textContent(field(message, "content"));
      }
    }
    throw new Error("LangChain result did not contain an assistant message");
  }
}

/** Build the production adapter from LangChain's official `createAgent` export. */
export function createLangChainReplyAgent(
  env: Record<string, string | undefined> = process.env,
): LangChainReplyAgent {
  const agent = createAgent({
    model: env.LANGCHAIN_MODEL ?? "openai:gpt-5.4",
    tools: [],
    systemPrompt: REPLY_INSTRUCTIONS,
  });
  return new LangChainReplyAgent(async (prompt) => {
    const result = await agent.invoke({ messages: [{ role: "user", content: prompt }] });
    return { messages: result.messages };
  });
}
